ref(tests) Try to improve consistency of store_event()

markstory opened the pull request:
We've hit a few scenarios where store_event() isn't resulting in a fully persisted event. Ensuring that the event can be fetched from eventstore should help eliminate a failure scenario.

markstory committed:
ref(tests) Try to improve consistency of store_event()

We've hit a few scenarios where store_event() isn't resulting in a fully
persisted event. Ensuring that the event can be fetched from eventstore
should help eliminate a failure scenario.

k-fish commented on src/sentry/testutils/cases.py:
```diff
@@ -678,6 +680,17 @@ def store_event(self, *args, **kwargs):
+                if did_store:
+                    break
+                attempt += 1
+                time.sleep(0.05)
```
`time.sleep(0.05 + attempt*0.05)` for a little backoff? 500ms might be too little total time? I think in terms of fixing flake I'd prefer we just fail the test if `store_event` fails to store the event in a timely matter, because that's the root of the type of flake, and we can look into reducing that time later.

markstory replied:
Failing the test is an option. I can try that out and see if it triggers often.

k-fish replied:
Yeah, I think if store_event never finishes, it would be a big red flag that we should look into. I'd rather fix it at the source then see it's effects indirectly through different layers of the test.

Zylphrex replied:
Would this cause problems with the visual snapshots and maybe other tests too?

For example, if a test is storing 2 or more events, the subsequent events that use `before_now` to set the timestamp will be pushed later, so the two events can be further apart in time than expected because of this sleep.

k-fish replied:
If that's the case, we can collect the ids and offer another `wait` method to be called later in the test after all the `store_event`'s have been called, which then would check that they have all been stored at once.

markstory replied:
> For example, if a test is storing 2 or more events, the subsequent events that use before_now to set the timestamp will be pushed later, so the two events can be further apart in time than expected because of this sleep.

Yes this would push time forward and could result in those events being farther apart than requested. We could not do this and instead have an assertion check that would let test authors wait for a provided event count, or fail. Something like

```python
self.store_event(...)
self.store_event(...)
# Wait for up to 1 second and then fail with an assertion error.
self.wait_for_event_count(2)
```

The 1 second is entirely arbitrary and I don't know yet what a good default value would be.

k-fish replied:
> The 1 second is entirely arbitrary and I don't know yet what a good default value would be. 

I would prefer waiting for the default selenium wait time, as if these events don't get stored then the rest of the test will like break, we can turn it down from there if it causes issues, but that would make it the most resilient to ensure the highest test pass rate when it's not going to fail.

markstory replied:
The default selenium wait time is ~30 seconds, however this method is also used in API test cases and the total failure scenario duration for those gets really high at 30 seconds per stored event.

k-fish replied:
Oh do we fail fast or collect failures? If we collect them then that's a fair point, I still think 1sec might be too low.

k-fish replied:
Regardless, we can work to make failure times around event storage better. If we have 30 second failures due to event storage on tests, then we should look into it and fix it.

markstory replied:
We don't fail fast in CI unfortunately.

markstory committed:
Fix failing test.

markstory committed:
Make wait a separate method

We don't want to invisibly add time to all tests. Instead we can use the
wait method on tests that are having timing challenges. Add the new wait
to an acceptance test that has been flaky lately.

Zylphrex commented on src/sentry/testutils/cases.py:
```diff
@@ -680,21 +680,28 @@ def store_event(self, *args, **kwargs):
+                break
+            attempt += 1
+            time.sleep(0.05)
+        if attempt == 10:
```
should this be using the `attempts` parameter?

markstory replied:
Indeed it should, my mistake.

markstory committed:
Fix mistake.
